chips/nrf52/src/spi.rs

```diff
@@ -0,0 +1,501 @@
+            // Pin select for MISO signal
+            // base + 0x510
+            pub psel_miso: VolatileCell<Pinmux>,
+            _reserved11: [u32; 4],
+            // SPI frequency
+            // base + 0x524
+            pub frequency: VolatileCell<Frequency>,
+            _reserved12: [u32; 3],
+            // Data pointer
+            // base + 0x534
+            pub rxd_ptr: VolatileCell<*mut u8>,
+            // Maximum number of bytes in receive buffer
+            // base + 0x538
+            pub rxd_maxcnt: VolatileCell<u32>,
+            // Number of bytes transferred in the last transaction
+            // base + 0x53C
+            pub rxd_amount: VolatileCell<u32>,
+            // EasyDMA list type
+            // base + 0x540
+            pub rxd_list: VolatileCell<u32>,
+            // Data pointer
+            // base + 0x544
+            pub txd_ptr: VolatileCell<*const u8>,
+            // Maximum number of bytes in transmit buffer
+            // base + 0x548
+            pub txd_maxcnt: VolatileCell<u32>,
+            // Number of bytes transferred in the last transaction
+            // base + 0x54C
+            pub txd_amount: VolatileCell<u32>,
+            // EasyDMA list type
+            // base + 0x550
+            pub txd_list: VolatileCell<u32>,
+            // Configuration register
+            // base + 0x554
+            pub config: VolatileCell<Config>,
+            _reserved13: [u32; 26],
+            // Over-read character. Character clocked out in case and over-read of the TXD buffer.
+            // base + 0x5C0
+            pub orc: VolatileCell<u32>,
+        }
+    }
+}
+
+
+// SPI-Master peripheral.
+pub struct SPIM {
+    registers: *const registers::spim::SPIM,
+    nvic_idx: NvicIdx,
+    client: Cell<Option<&'static hil::spi::SpiMasterClient>>,
+    chip_select: Cell<Option<&'static hil::gpio::Pin>>,
+    initialized: Cell<bool>,
+    busy: Cell<bool>,
+    tx_buf: TakeCell<'static, [u8]>,
+    rx_buf: TakeCell<'static, [u8]>,
+    transfer_len: Cell<usize>,
+}
+
+impl SPIM {
+    pub const fn new(instance: usize) -> SPIM {
+        SPIM {
+            registers: registers::spim::INSTANCES[instance].0,
+            nvic_idx: registers::spim::INSTANCES[instance].1,
+            client: Cell::new(None),
+            chip_select: Cell::new(None),
+            initialized: Cell::new(false),
+            busy: Cell::new(false),
+            tx_buf: TakeCell::empty(),
+            rx_buf: TakeCell::empty(),
+            transfer_len: Cell::new(0),
+        }
+    }
+
+    fn regs(&self) -> &registers::spim::SPIM {
+        unsafe { &*self.registers }
+    }
+
+    #[inline(never)]
+    pub fn handle_interrupt(&self) {
+
+        if self.regs().events_end.get() == 1 {
+            // End of RXD buffer and TXD buffer reached
+            match self.chip_select.get() {
+                Some(cs) => cs.set(),
+                None => panic!("Chip select not configured."),
```

Comment:
In general, we try to avoid `panic`ing if we can help it -- on the downcall you returned `ENODEVICE` which makes a lot of sense. It's not immediately obvious to me how you would even reach the callback w/out having set a chip select (in practice) here, so maybe it's okay.

Perhaps just adding a comment here that explains when this `panic` (should) never trigger in practice would alleviate my concerns. Perhaps it makes sense to convert this to a `debug_assert` and just bail out in the production case. Unsure..